### ISSUE TITLE
Prefer string.Contains() over string.IndexOf() != -1

### ISSUE THREAD

**Issue by terrajobst** [dotnet/runtime#33785]
Calls to `String.IndexOf(...)` where the result is then just compared to -1 can instead be replaced by calls to `String.Contains(...)`.

**Category**: Performance

**Comment by jeffhandley** [comment]
Estimates:
* Analyzer: Small
* Fixer: Small

**Comment by stephentoub** [comment]
We actually need to be a bit careful on this one.  For whatever reason, the semantics of string.IndexOf vs string.Contains differ, in that the former does a culture-aware comparison and the latter does an ordinal comparison.

**Comment by terrajobst** [comment]
We discussed this today and concluded we should offer two fixers:

* Replace with `String.Contains` and use ordinal
* Replace with `String.Contains` and preserve culture sensitive comparison

(wording to be defined)

**Comment by pgovind** [comment]
Taking a crack at this one

**Comment by pgovind** [comment]
Just adding a note here that `string.Contains(char, StringComparison)` just calls through to `string.IndexOf(char, StringComparison)`, which is what we're trying to fix in the first place. So, we should only offer `string.Contains(char)` as the code-fix for `string.IndexOf(char)`. (https://source.dot.net/#System.Private.CoreLib/String.Searching.cs,32)

**Comment by pgovind** [comment]
Also, as @danmosemsft asks [here](https://github.com/dotnet/runtime/pull/35350#issuecomment-618709868), should we also tackle `string.IndexOf(string/char, [StringComparison.Ordinal]) >= 0` -> `string.Contains(string/char)` in this analyzer/code-fix? I think that's a good idea.

**Comment by GrabYourPitchforks** [comment]
There's some amount of indirection in the `string.IndexOf(string)` code paths compared to `string.Contains(string)`, but we've been working to reduce it. The fixed-size overhead shouldn't be more than a few cycles.

The bigger benefit IMO is code clarity. Which of these is more understandable?

```cs
if (theString.Contains("hello")) { /* do something */ }
if (theString.IndexOf("hello") != -1) { /* do something */ }
if (theString.IndexOf("hello") != 1) { /* do something */ } // <-- did you spot the typo?
if (theString.IndexOf("hello") >= 0) { /* do something */ }
```

There is also as Steve mentioned the fact that `string.Contains` and `string.IndexOf` use different comparisons by default. I can easily justify why we should use one or the other comparison as our One True Default™. But it's truly maddening that we mixed and matched on the `string` class. Even worse, we mixed and matched on overloads of the same function: `string.IndexOf(char)` uses a different default comparison than `string.IndexOf(string)`!

**Comment by jeffhandley** [comment]
I agree with the clarity/readability argument. To go that route here, I think we would want to always recommend `Contains` even when there is a negligible perf impact. And that introduces risk for these overloads having different comparisons.

I definitely don’t think we should put this into the “performance” category of analyzers. And with the inconsistencies we have, I’m worried about subtle regressions we could lead developers to introduce with this analyzer.

@pgovind - what do you recommend we do here?

**Comment by GrabYourPitchforks** [comment]
Related: https://github.com/dotnet/roslyn-analyzers/issues/2581#issuecomment-584438937, which suggests an analyzer that says "You should pass an explicit `StringComparison` into these APIs (where the default behavior isn't `Ordinal`)."

Then you get the following order of operations when pairing that analyzer with this one. All three lines below have the exact same behavior.

```cs
// Original code
bool exists = theString.IndexOf("target") >= 0;

// First apply CA1307
// (assuming caller wanted culture-aware behavior)
bool exists = theString.IndexOf("target", StringComparison.CurrentCulture) >= 0;

// Then apply the analyzer as proposed here
bool exists = theString.Contains("target", StringComparison.CurrentCulture);
```

**Comment by pgovind** [comment]
Thanks for the explanation Levi. I'm going to go ahead and change the analyzer to appropriately convert `string.IndexOf`(when the result is compared to 1/-1) to `string.Contains` always in favor of readability(and we take the slight hit in performance because `string.Contains(StringComparison)` just calls through to `string.IndexOf`.)

**Comment by stephentoub** [comment]
> I'm going to go ahead and change the analyzer to appropriately convert string.IndexOf(when the result is compared to 1/-1) to string.Contains always in favor of readability(and we take the slight hit in performance because string.Contains(StringComparison) just calls through to string.IndexOf.)

How much is the perf hit?

**Comment by GrabYourPitchforks** [comment]
> How much is the perf hit?

Currently `string.Contains(string, StringComparison)` is identical to `string.IndexOf(string, StringComparison) >= 0`. So it should have identical perf since presumably the latter is what the dev had already written in the first place.

Separately, I was experimenting with making the JIT smarter about these APIs, basically saying "if you see a call to `string.IndexOf`, _and_ if the second parameter is the constant `StringComparison.Ordinal`, then pretend the call site was for this other method `string.InternalIndexOfOrdinal` instead." Was looking at it in the same vein of optimizing `if (someString == string.Empty) { /* ... */ }` patterns, where the JIT would detect that one of the arguments was the value `string.Empty` and would call a more optimized method instead.

Didn't get too far into the investigation because I ran into JIT layering issues. And didn't pursue it more aggressively because I lack evidence that these are bottlenecks. But if we have that evidence then that might resurrect things. :)

**Comment by jeffhandley** [comment]
In hindsight, our estimation of this was off -- there have been a lot more edge cases and scenarios that needed attention then we had anticipated.  The more accurate T-shirt estimates would have been:

* Analyzer: Medium
* Fixer: Small

**Comment by danmosemsft** [comment]
linking https://github.com/dotnet/roslyn-analyzers/pull/3610

**Comment by pgovind** [comment]
Fixed with dotnet/roslyn-analyzers#3610
